cltk/phonology/utils.py

```diff
@@ -0,0 +1,497 @@
+        else:
+            re_before = r"(?<=["
+            for phoneme in phonology:
+                if phoneme.match_list(self.position.before):
+                    re_before += phoneme.ipar
+            re_before += r"])"
+
+        if self.position.position == "last":
+            re_after = r"$"
+        elif self.position.after is None:
+            re_after = r""
+        else:
+            re_after = r"(?=["
+            for phoneme in phonology:
+                if phoneme.match_list(self.position.after):
+                    re_after += phoneme.ipar
+            re_after += "])"
+        return re_before+self.temp_sound.ipar+re_after
+
+    @staticmethod
+    def from_regular_expression(re_rule, estimated_sound, ipa_class):
+        """
+
+        :param re_rule: pattern (first argument of re.sub)
+        :param estimated_sound: an IPA character (second argument of re.sub)
+        :param ipa_class: dict whose keys are IPA characters and values are Vowel or Consonant instances
+        :return: corresponding Rule instance
+        """
+        assert len(re_rule) > 0
+        if re_rule[0] == "^":
+            place = "first"
+        elif re_rule[-1] == "$":
+            place = "last"
+        else:
+            place = "inner"
+
+        before_pattern = r"(?<=\(\?\<\=\[)\w*"
+        core_pattern = r"(?<=\))\w(?=\(\?\=)|(?<=\^)\w(?=\(\?\=)|(?<=\))\w(?=\$)"
+        after_pattern = r"(?<=\(\?\=\[)\w*"
+        before_search = re.search(before_pattern, re_rule)
+        core_search = re.search(core_pattern, re_rule)
+        after_search = re.search(after_pattern, re_rule)
+        if before_search is None:
+            before = None
+        else:
+            before = [ipa_class[ipar].to_abstract() for ipar in before_search.group(0)]
+        if core_search is not None:
+            core = ipa_class[core_search.group(0)]
+        else:
+            logger.error("No core")
+            raise ValueError
+        if after_search is None:
+            after = None
+        else:
+            after = [ipa_class[ipar].to_abstract() for ipar in after_search.group(0)]
+        abstract_position = AbstractPosition(place, before, after)
+        return Rule(abstract_position, core, ipa_class[estimated_sound])
+
+    def __add__(self, other):
+        assert isinstance(other, Rule)
+        assert self.position.same_place(other.position)
+        assert self.temp_sound.ipar == other.temp_sound.ipar
+        assert self.estimated_sound.ipar == other.estimated_sound.ipar
+        position = self.position + other.position
+        return Rule(position, self.temp_sound, self.estimated_sound)
+
+
+class Transcriber:
+    """
+    There are two steps to transcribe words:
+        - firstly, a greedy approximation of the pronunciation of word
+        - then, use of rules to precise pronunciation of a preprocessed list of transcribed words
+    """
+    def __init__(self, diphthongs_ipa, diphthongs_ipa_class, ipa_class, rules):
```

Comment:
You may want to further parametarize this, in case you'll need a triphthong class in the near future.